notarization_poller/src/notarization_poller/worker.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python
+"""Notarization poller worker functions.
+
+Attributes:
+    log (logging.Logger): the log object for the module.
+
+"""
+import asyncio
+import logging
+import signal
+import socket
+import sys
+import typing
+from asyncio import sleep
+
+import aiohttp
+import arrow
+from taskcluster.aio import Queue
+
+from notarization_poller.config import get_config_from_cmdln, update_logging_config
+from notarization_poller.task import Task, claim_work
+from scriptworker_client.constants import STATUSES
+from scriptworker_client.utils import makedirs, rm
+
+log = logging.getLogger(__name__)
+
+
+# RunTasks {{{1
+class RunTasks:
+    """Manages processing of Taskcluster tasks."""
+
+    def __init__(self, config):
+        """Initialize RunTasks."""
+        self.config = config
+        self.running_tasks = []
+        self.last_claim_work = arrow.get(0)
+        self.is_cancelled = False
+        self.future = None
+
+    async def invoke(self):
+        """Claims and processes Taskcluster work."""
+        while not self.is_cancelled:
+            num_tasks_to_claim = self.config["max_concurrent_tasks"] - len(self.running_tasks)
+            if num_tasks_to_claim > 0 and arrow.utcnow().timestamp - self.last_claim_work.timestamp >= self.config["claim_work_interval"]:
+                async with aiohttp.ClientSession() as session:
+                    queue = Queue(
+                        options={
+                            "credentials": {"accessToken": self.config["taskcluster_access_token"], "clientId": self.config["taskcluster_client_id"]},
+                            "rootUrl": self.config["taskcluster_root_url"],
+                        },
+                        session=session,
+                    )
+                    new_tasks = await self._run_cancellable(claim_work(self.config, queue, num_tasks=num_tasks_to_claim))
+                self.last_claim_work = arrow.utcnow()
+                for claim_task in new_tasks.get("tasks", []):
+                    new_task = Task(self.config, claim_task)
+                    self.running_tasks.append(new_task)
+            await self.prune_running_tasks()
+            sleep_time = self.last_claim_work.timestamp + self.config["claim_work_interval"] - arrow.utcnow().timestamp
+            sleep_time > 0 and await self._run_cancellable(sleep(sleep_time))
```

Comment:
I *think* if the body of this functions is wrapped in an `asyncio.Task`, then you don't need `_run_cancellable`, and cancelling the task will propagate to the awaited futures here.

Reply:
I think I've done this in the latest working patchset.

Reply:
I still see `_run_cancellable` here. It exists to propagate cancellation to the `sleep` and `claim_work` calls. If you don't have that function, and cancel the future from `invoke`, it should automatically propagate the cancellation to the `sleep` and `claim_work` calls.

That said, this seems like something that can be left for a followup.